test/k8s-integration/version_test.go

```diff
@@ -0,0 +1,309 @@
+package main
+
+import (
+	"fmt"
+	"testing"
+)
+
+func TestParseVersion(t *testing.T) {
+	tests := []struct {
+		version   string
+		expectErr bool
+		expectedV version
+	}{
+		// Positive test cases.
+		{
+			version: "v1.1.1",
+			expectedV: version{
+				version: [4]int{1, 1, 1, -1},
+			},
+		},
+		{
+			version: "v1.18.0",
+			expectedV: version{
+				version: [4]int{1, 18, 0, -1},
+			},
+		},
+		{
+			version: "v1.18.0-gke.0",
+			expectedV: version{
+				version: [4]int{1, 18, 0, 0},
+			},
+		},
+		{
+			version: "1.18.3-gke.10",
+			expectedV: version{
+				version: [4]int{1, 18, 3, 10},
+			},
+		},
+		{
+			version: "1.18.9",
+			expectedV: version{
+				version: [4]int{1, 18, 9, -1},
+			},
+		},
+		{
+			version: "1.18.10-gke.10",
+			expectedV: version{
+				version: [4]int{1, 18, 10, 10},
+			},
+		},
+		{
+			version: "10.18.10-gke.10",
+			expectedV: version{
+				version: [4]int{10, 18, 10, 10},
+			},
+		},
+		{
+			version: "100.101.102-gke.103",
+			expectedV: version{
+				version: [4]int{100, 101, 102, 103},
+			},
+		},
+		// Negative test cases
+		{
+			version:   "1",
+			expectErr: true,
+		},
+		{
+			version:   "-1.18.9",
+			expectErr: true,
+		},
+		{
+			version:   "1.-18.9",
+			expectErr: true,
+		},
+		{
+			version:   "1.18.-9",
+			expectErr: true,
+		},
+		{
+			version:   "1.18.9-gke.-1",
+			expectErr: true,
+		},
+		{
+			version:   "1.1",
+			expectErr: true,
+		},
+		{
+			version:   "1.18",
+			expectErr: true,
+		},
+		{
+			version:   "1.18-gke.0",
+			expectErr: true,
+		},
+		{
+			version:   "1.18.0-alpha.x",
+			expectErr: true,
+		},
+		{
+			version:   "1.18.0-alpha.beta.1",
+			expectErr: true,
+		},
+		{
+			version:   "alpha.3.673+73326ef01d2d7c",
+			expectErr: true,
+		},
+		{
+			version:   "1.18-alpha.3.673+73326ef01d2d7c",
+			expectErr: true,
+		},
+		{
+			version:   "1.18.3-alpha.3.673+73326ef01d2d7c",
+			expectErr: true,
+		},
+	}
+
+	fmt.Println("Num test cases:", len(tests))
+	for i, tc := range tests {
+		fmt.Printf("Run test: %s, test case number:%d\n", t.Name(), i)
+		gotV, err := parseVersion(tc.version)
+		if err != nil {
+			if !tc.expectErr {
+				t.Fatalf("Got unexpected err: %v", err)
+			}
+			continue
+		}
+
+		if err == nil && tc.expectErr {
+			t.Fatal("Got no error but expected one")
+			continue
+		}
+
+		if gotV.version[0] != tc.expectedV.version[0] ||
+			gotV.version[1] != tc.expectedV.version[1] ||
+			gotV.version[2] != tc.expectedV.version[2] ||
+			gotV.version[3] != tc.expectedV.version[3] {
+			t.Fatalf("Got version: %s, expected: %s", gotV.String(), tc.expectedV.String())
```

Comment:
Adding the test index might make tracking down errors easier if there's ambiguity in the expected/got output.

Reply:
A test pattern I've seen is to use `t.Run` and which gives each test case a descriptive name. Then you also don't have the issue where a `Fatalf` in one test case aborts the remaining test cases